### DISCUSSION TITLE
How to save charts to saparate .png files?

### DISCUSSION THREAD

**Discussion by k605543016** [PHPOffice/PhpSpreadsheet#2656] [Q&A] [answered]
I'm using phpSpreadSheet class to successfully create xlsx spreadsheet with several charts.The spreadsheet is wrtitten down to disk. After that I would like to export these charts from spreadsheet to seperate .png files. Based on documentation I tried to use the following code:

```
$inputFileType = \PhpOffice\PhpSpreadsheet\IOFactory::identify($XLSXfilepath);
$reader = IOFactory::createReader($inputFileType);
$reader->setIncludeCharts(true);
$spreadsheet = $reader->load($XLSXfilepath, $reader::LOAD_WITH_CHARTS);

$nr = 1;
foreach ($spreadsheet->getWorksheetIterator() as $worksheet) {
    $chartNames = $worksheet->getChartNames();
    foreach ($chartNames as $chartName) {
        $chart = $worksheet->getChartByName($chartName);
        try{
            //the following code results false: 
            $res = $chart->render('temp/chart'.$nr.'.png');
        } catch(Exception $e){
            //this doesn't happen, which means then no exception is thrown:
            print $e->getMessage();
        }
        $nr++;
    }
}
```

The problem is that $chart->render results false with any error message and - of course - no .png file is saved on disk. But no exception is thrown. I double-checked that I have full rights to save file to selected location and of course there is enough free space on disk... I also checked that foreach loop properly finds all charts, which are saved in .xlsx file.

Is there any way to find out why the charts are not rendered to .png files?

**Comment by MarkBaker** [comment, marked as answer]
First check: have you installed jpgraph and configured that as your chart renderer?

**Reply by k605543016** [reply]
Yes. 
Is it required to use jpgraph if I try to render charts to .png files? 
If I do it, charts look a little different trom those generated by Excel...

**Reply by MarkBaker** [reply]
You need something to create that png from the cart data: we chose to use jpgraph as a library to do this, and there aren't really any alternatives (even today, there are few PHP libraries for rendering chart data as static image files; most libraries seem to use js for animated charts instead). While jpgraph is far from perfect, it supported most of the chart types that are available in MS Excel, with the functionality to render them as an image file.

**Reply by k605543016** [reply]
OK - I understand - so there is no way to just "extract" (or copy) the chart, which was created originally by Excel and saved inside spreadsheet, as a separate .png file? It should be "recreated" using jpgraph... Am I right?

**Reply by MarkBaker** [reply]
It is a PHP application, not MS Excel itself. The chart doesn't exist as an image in the Excel file, but as a series of property values that define how it's structured, what type of chart it is, whether it displays a key or not, what the axes are, etc. We simply try and create a similar definition that jpgraph can recognise to render.

Even different spreadsheet programs will show some differences when rendering a chart from the same file:
| ![image](https://user-images.githubusercontent.com/770298/156941625-b4ad10d6-eb86-4d92-8281-10bae72ba39c.png)
|:--:|
| MS Excel

| ![image](https://user-images.githubusercontent.com/770298/156941644-fb62b4aa-510e-43b9-b081-dedb441e91a6.png)
|:--:|
| Libre Office

| ![image](https://user-images.githubusercontent.com/770298/156941663-86106061-8889-4eda-b35e-028bccf8f64f.png)
|:--:|
| Gnumeric

**Reply by MarkBaker** [reply]
While we try to create something similar using jpgraph, including the complexities  of multiple charts of different types in the same plot area, there's limits to what we can do, especially as jpgraph is no longer actively maintained as an OS library.
